### PR TITLE
Add exploration strategies

### PR DESCRIPTION
**Pull request by Yokiball**

### CODE REVIEW

File: ale/teacher/hybrid/sequential_representation_lc.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+from ale.trainer.predictor import Predictor
+from ale.teacher.exploration.k_means import ClusteredDocuments, ClusterDocument
+from ale.trainer.prediction_result import TokenConfidence, PredictionResult
+
+class NGramVectors:
+    def __init__(self, sizes: List[int], seed: int, lexical_dimension: int = 50, word_embedding_dimension: int = 50) -> None:
+        self.sizes: List[int] = sizes
+        self.seed: int = seed
+        self.lexical_dimension: int = lexical_dimension
+        self.word_embedding_dimension: int = word_embedding_dimension
```
Comment (philipp-kohl) [inline ale/teacher/hybrid/sequential_representation_lc.py]: Do you use this var?

File: ale/teacher/hybrid/sequential_representation_lc.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        self.vector_dict: Dict[str,np.ndarray] = dict()
+
+    def generate_n_grams(self, token: str) -> List[np.ndarray]:
```
Comment (philipp-kohl) [inline ale/teacher/hybrid/sequential_representation_lc.py]: generate_lexical_token_vector?

File: ale/teacher/hybrid/sequential_representation_lc.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        for size in self.sizes:
+            subsequences_of_size: List[str] = [token[i:i+size] for i in range(len(token)-size)]
+            subsequences.extend(subsequences_of_size)
+        for seq in subsequences:
+            seq_vector = self.vector_dict.get(seq)
+            if not seq_vector: # vector n gram does not exist yet
+                np.random.seed(self.seed) # use ALE seed
+                self.vector_dict[seq] = np.random.random(size=(self.lexical_dimension)) # lexical vector of dimension given, per default 50
+            vectors.append(self.vector_dict[seq])
+        return vectors
```
Comment (philipp-kohl) [inline ale/teacher/hybrid/sequential_representation_lc.py:54]: Do we need the vectors at this point? The first step would be to create a mapping of each n-gram to lexical vector?

File: ale/teacher/hybrid/sequential_representation_lc.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        vector_sum: np.ndarray = sum(vectors)
+        norm = np.linalg.norm(vector_sum)
+        if norm == 0:
+            return vector_sum
+        return vector_sum/norm
+    
+def get_word2vec_skipgram_vector(token: str, word2vec: KeyedVectors) -> np.ndarray:
+    """ Generates word2vec (skip gram) vectors with gensim
+    """
+    return word2vec[token]
```
Comment (philipp-kohl) [inline ale/teacher/hybrid/sequential_representation_lc.py]: intention of this function?

File: ale/teacher/hybrid/sequential_representation_lc.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+def cluster_documents(corpus: Corpus, k: int, embeddings: Dict[int,np.ndarray]) -> ClusteredDocuments:
+    model = KMeans(n_clusters=k, init='k-means++',
+                    max_iter=300, n_init='auto')
+    X = list(embeddings.values())
+    model.fit(X)
+
+    # get the distance to the corresponding cluster centroid for each document
+    centers = model.cluster_centers_
+    clustered_documents: List[ClusteredDocuments] = []
+    for i in range(len(list(embeddings.keys()))):
```
Comment (philipp-kohl) [inline ale/teacher/hybrid/sequential_representation_lc.py]: why no enumerate?

File: ale/teacher/hybrid/sequential_representation_lc.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import random
+import numpy as np
+from gensim.models import Word2Vec,KeyedVectors
+from sklearn.cluster import KMeans
+from ale.config import NLPTask
+from ale.corpus.corpus import Corpus
+from ale.registry.registerable_teacher import TeacherRegistry
+from ale.teacher.base_teacher import BaseTeacher
+from ale.trainer.predictor import Predictor
+from ale.teacher.exploration.k_means import ClusteredDocuments, ClusterDocument
```
Comment (philipp-kohl) [inline ale/teacher/hybrid/sequential_representation_lc.py]: Move models to own file?

**Review by philipp-kohl — COMMENTED** [review]
WIP

File: ale/teacher/hybrid/sequential_representation_lc.py (hunk 3 above)
Comment (Yokiball) [reply ale/teacher/hybrid/sequential_representation_lc.py:54]: Yes, I thought we do this iteratively. So for each token, this function checks if the lexical vector exists in the mapping. If not, it generates and then returns them. Otherwise, it returns them directly.

File: ale/teacher/exploitation/fluctuation_historical_sequence.py [hunk 7]
```diff
@@ -11,32 +11,30 @@
...
         """
         self.idx: int = idx
         self.historical_confidences: List[float] = [confidence]
+        self.window_size: int = window_size
 
-    def update_confidences(self, confidence: float, window_size: int) -> List[float]:
+    def update_confidences(self, confidence: float) -> List[float]:
         """
-        Updates confidence scores: If window-size confidences are saved, removes oldest cycles and saves new one. Else, adds new scores.
+        Appends the new confidence scores to the historical confidence list.
```
Comment (Yokiball) [inline ale/teacher/exploitation/fluctuation_historical_sequence.py:24]: Gibt es einen Grund, wieso wir die alle speichern wollen statt nur das Window?

File: ale/teacher/exploitation/fluctuation_historical_sequence.py [hunk 8]
```diff
@@ -110,68 +103,76 @@ def __init__(
...
+        out_ids: List[int] = self.compute_function(prediction_results, step_size)
+
+        return out_ids
+
+    def compute_ner(self, predictions: Dict[int, PredictionResult], step_size: int) -> List[int]:
+        """
+        Fluctuation of Historical Sequence is calculated on instance-level,
+        LC is used with token-score, aggregated as configured.
+        """
+        if self.is_initial_train():  # TODO Question: Do we have a model at this point?
```
Comment (Yokiball) [inline ale/teacher/exploitation/fluctuation_historical_sequence.py]: Frage geklärt?

File: ale/teacher/exploitation/fluctuation_historical_sequence.py [hunk 9]
```diff
@@ -110,68 +103,76 @@ def __init__(
...
+            next_batch = random.sample(list(predictions.keys()), step_size)
+            instance_scores: Dict[int, float] = self.compute_least_confidence_instance_scores(predictions)
+            self.historical_sequence: HistoricalSequence = HistoricalSequence(instance_scores,
+                                                                              self.historical_window_size)
+            return next_batch  # TODO Question: returning the random batch for comparison reason?
```
Comment (Yokiball) [inline ale/teacher/exploitation/fluctuation_historical_sequence.py]: Frage geklärt?

File: ale/teacher/exploitation/fluctuation_historical_sequence.py [hunk 10]
```diff
@@ -110,68 +103,76 @@ def __init__(
...
         return out_ids
 
+    def compute_fluctuation_scores_per_doc(self, historical_sequences) -> Dict[int, float]:
+        fluctuation_scores: Dict[int, float] = dict()
+        for idx, hist_scores in historical_sequences.items():
+            fluctuation_scores[idx] = self.compute_fhs_score(hist_scores)
+        return fluctuation_scores
+
+    def compute_fhs_score(self, hist_scores: List[float]) -> float:
+        weight = self.weight  # in paper (11) referred as w_s
```
Comment (Yokiball) [inline ale/teacher/exploitation/fluctuation_historical_sequence.py]: Es muss glaube ich "referred to as w_s" heißen?

File: ale/teacher/exploitation/fluctuation_historical_sequence.py [hunk 11]
```diff
@@ -110,68 +103,76 @@ def __init__(
...
+        return fluctuation
+
     def compute_least_confidence_instance_scores(self, predictions: Dict[int, PredictionResult]) -> Dict[int, float]:
         scores = dict()
         for idx, prediction in predictions.items():
             token_confidences: List[TokenConfidence] = prediction.ner_confidences_token
             highest_confidences: List[float] = []
             for token in token_confidences:
-                # LC on token-level, reversed to take max
+                # LC on token-level, reversed to take max  # TODO Question: Why not taking MIN?
```
Comment (Yokiball) [inline ale/teacher/exploitation/fluctuation_historical_sequence.py]: Ich hatte mich da am Paper orientiert, können wir aber auch ändern

File: ale/teacher/exploitation/max_tag_count.py [hunk 12]
```diff
@@ -23,13 +23,13 @@ class MaxTagCountTeacher(BaseTeacher, ABC):
...
-        labels: List[any],
-        nlp_task: NLPTask,
-        aggregation_method: Optional[AggregationMethod]
+            self,
+            corpus: Corpus,
+            predictor: Predictor,
+            seed: int,
+            labels: List[any],
+            nlp_task: NLPTask,
+            aggregation_method: Optional[AggregationMethod] = AggregationMethod.SUM
```
Comment (Yokiball) [inline ale/teacher/exploitation/max_tag_count.py:34]: Wieso ist SUM der default? Sollten wir vllt dazu schreiben

File: ale/teacher/exploitation/tag_flip_historical.py [hunk 13]
```diff
@@ -16,37 +16,53 @@ def __init__(self, idx: int, prediction_result: PredictionResult) -> None:
...
+
+    def compute_diffs(self) -> List[int]:
+        """
+        Compares old and new labels and returns a list indicating which label predictions have changed (per token).
+
         Returns:
-            - difference (List[int]): a list which indicates if the token-predictions has changed (1) or not (0)
+            - difference (List[int]): a list which indicates how often the token-predictions has changed over time
         """
+        if len(self.labels) == 1:
```
Comment (Yokiball) [inline ale/teacher/exploitation/tag_flip_historical.py]: hmmm, wann kann das vorkommen? Haben wir nicht immer mind. 2? Also 1 Label und '0'?

File: ale/teacher/exploitation/tag_flip_historical.py [hunk 14]
```diff
@@ -55,37 +71,36 @@ def __init__(self, docs_with_predictions: Dict[int, PredictionResult]) -> None:
...
-                    doc, docs_with_predictions[doc])
-                # doc never predicted before, so no difference
-                difference: List[int] = [
-                    0 for item in historical_doc.get_labels()]
-                differences[doc] = difference
+        for doc_idx, prediction in docs_with_predictions.items():
+            differences[doc_idx] = self.compute_difference_for_doc(doc_idx, prediction)
         return differences
 
+    def compute_difference_for_doc(self, doc_idx: int, prediction: PredictionResult):
```
Comment (Yokiball) [inline ale/teacher/exploitation/tag_flip_historical.py]: Finde den Namen der Methode ein bisschen irreführend. Vielleicht update_and_compute_difference_for_doc? Oder update_history einzeln und dann compute_diffs?

File: ale/teacher/exploration/k_means.py [hunk 15]
```diff
@@ -100,45 +78,47 @@ def __init__(self, corpus: Corpus, predictor: Predictor, seed: int, labels: List
...
-                    if len(potential_docs_cluster) < docs_per_rest_clusters: # less docs in cluster left than needed
-                        output_ids.extend(potential_docs_cluster)
-                        empty_clusters.append(cluster)
-                    else:
-                        output_ids.extend(potential_docs_cluster[:docs_per_rest_clusters])
-        
-        out_ids = [item.idx for item in output_ids]
-        return out_ids
+    def propose(self, potential_ids: List[int], step_size: int, budget: int) -> List[int]:
+        pass#return propose_nearest_neighbors_to_centroids(self.clustered_documents, potential_ids, step_size, budget)
```
Comment (Yokiball) [inline ale/teacher/exploration/k_means.py]: Wieso pass?

File: ale/teacher/exploration/utils/cluster_helper.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+        return [self.id2doc[doc_id] for doc_id in indices]
+
+    def get_docs_by_cluster_idx(self, cluster_idx: int) -> List[ClusterDocument]:
+        return self.cluster2doc[cluster_idx]
+
+    def get_potential_docs_by_cluster_idx(self, cluster_idx: int, potential_ids: List[int]) -> List[ClusterDocument]:
+        return [d for d in self.cluster2doc[cluster_idx] if d.idx in potential_ids]
+
+
+class ClusterHelper:
```
Comment (Yokiball) [inline ale/teacher/exploration/utils/cluster_helper.py:45]: I like :)

**Review by Yokiball — COMMENTED** [review]
Sieht sehr gut aus, hab nur Kleinigkeiten angemerkt :)

File: ale/teacher/exploitation/fluctuation_historical_sequence.py (hunk 7 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/fluctuation_historical_sequence.py:24]: This provides more flexibility to test whether it is beneficial to make the window larger or shorter. It allows even dynamic approaches. The memory footprint should not be that high.

File: ale/teacher/exploitation/fluctuation_historical_sequence.py (hunk 8 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/fluctuation_historical_sequence.py]: Yes, I will rename the is_initial_train  to is_first_call

File: ale/teacher/exploitation/fluctuation_historical_sequence.py (hunk 9 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/fluctuation_historical_sequence.py]: Yes

File: ale/teacher/exploitation/fluctuation_historical_sequence.py (hunk 10 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/fluctuation_historical_sequence.py]: You are right

File: ale/teacher/exploitation/fluctuation_historical_sequence.py (hunk 11 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/fluctuation_historical_sequence.py]: Sounds good

File: ale/teacher/exploitation/max_tag_count.py (hunk 12 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/max_tag_count.py:34]: Added a comment

File: ale/teacher/exploitation/tag_flip_historical.py (hunk 13 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/tag_flip_historical.py]: That means that we only have a single prediction in the cache. We cannot determine a tag flip on a single prediction. Therefore just return 0, representing no flips detected. But I think I will rename the variable because it seems to be misleading.

Comment (philipp-kohl) [reply ale/teacher/exploitation/tag_flip_historical.py]: I renamed it to predictions

File: ale/teacher/exploitation/tag_flip_historical.py (hunk 14 above)
Comment (philipp-kohl) [reply ale/teacher/exploitation/tag_flip_historical.py]: Good point! I changed the method to update only the historical_doc and return it. Then, compute the differences in the for loop. Please take a look at it again.

File: ale/teacher/exploration/k_means.py (hunk 15 above)
Comment (philipp-kohl) [reply ale/teacher/exploration/k_means.py]: Unused strategy for the moment. Did not apply the refactoring to it. I will take a look at it now.

Comment (philipp-kohl) [reply ale/teacher/exploration/k_means.py]: Refactoring done. Please take a look at it.

File: ale/teacher/exploitation/tag_flip_historical.py (hunk 14 above)
Comment (Yokiball) [reply ale/teacher/exploitation/tag_flip_historical.py]: Yes, looks good now

File: ale/teacher/exploration/k_means.py (hunk 15 above)
Comment (Yokiball) [reply ale/teacher/exploration/k_means.py]: Looks good now
